File: 2ndSemester/AlgorithimsAndDataStructuresII/U07-BinaryTrees/c/binarytree.c
#include <err.h>
#include <stdlib.h>
#include <stdio.h>
#include "binarytree.h"

Node* root;

void start(){
  root = NULL;
}

void insert(int x){
  root = insertRec(x, root);
} 

Node* insertRec(int x, Node *n){
  if(n == NULL)
    n = newNode(x);
  else if(x < n->element)
    n->left = insertRec(x, n->left);
  else if(x > n->element)
    n->right = insertRec(x, n->right);
  else
    errx(1, "Element %d is already on your tree", x);
  return n;
}

void insertByFather(int x){
  if(root == NULL)
    root  = newNode(x);
  else if(x < root->element)
    insertByFatherRec(x, root->left, root);
  else if(x > root->element)
    insertByFatherRec(x, root->right, root);
  else 
    errx(1, "Element %d is already on your tree", x);
}

void insertByFatherRec(int x, Node *i, Node *father){
  if(i == NULL){
    if(x < i->element)
      father->left = newNode(x);
    else
      father->right = newNode(x); 
  }
  else if(x < i->element)
    insertByFatherRec(x, i->left, i);
  else if(x > i->element)
    insertByFatherRec(x, i->right, i);
  else 
    errx(1, "Element %d is already on your tree", x);
}

void removeNode(int key){
  root = removeRec(key, root);
}

Node* removeRec(int key, Node *n){
  if(n == NULL)
    errx(1, "The tree is empty");
  else if(key < n->element)
    n->left = removeRec(key, n->left);
  else if(key > n->element)
    n->right = removeRec(key, n->right);
  else if(n->right == NULL)
    n = n->left;
  else if(n->left == NULL)
    n = n->right;
  else
    n->left = getAntecessor(n, n->left);
  return n;
}

Node* getAntecessor(Node *i, Node *j){
  if(j->right == NULL){
    i->element = j->element;
    j = j->left;
  }
  else 
    j = getAntecessor(i, j->right);
  return j;
}

void displayCentral(){
  displayCentralRec(root);
}

void displayCentralRec(Node *n){
  if (n != NULL){
    displayCentralRec(n->left);
    printf("%d\n", n->element);
    displayCentralRec(n->right);
  } 
}

void displayBefore(){
  displayBefore(root);
}

void displayBeforeRec(Node *n){
  if(n != NULL){
    printf("%d\n", n->element);
    displayBeforeRec(n->left);
    displayBeforeRec(n->right);
  }
}

void displayAfter(){
  displayAfterRec(root);
} 

void displayAfterRec(Node* n){
  if(n != NULL){
    displayAfterRec(n->left);
    displayAfterRec(n->right);
    printf("%d\n", n->element);
  }
}

File: 2ndSemester/AlgorithimsAndDataStructuresII/U08-BalancedBinaryTrees/Basic Implementation/BinaryTree.java
public class BinaryTree{
  public Node root;
  public BinaryTree(){
    this.root = null;
  }
  
  public void insert(int el){
    root = insert(root, el);
  }

  private Node insert(Node n, int el){
    if(n == null)
      n = new Node(el);
    else if(el < n.element)
      n.left = insert(n.left, el);
    else if(el > n.element)
      n.right = insert(n.right, el);
    else
      System.out.println("The element " + el + " already exists inside your tree");
    return n;
  }

  public boolean search(int key){
    return search(root, key);
  }
  
  private boolean search(Node n, int key){
    boolean foundKey;
    if(n == null)
      foundKey = false;
    else if(n.element == key)
      foundKey = true;
    else if(key < n.element)
      foundKey = search(n.left, key);
    else
      foundKey = search(n.right, key);
    return foundKey;
  }

  public void displayInOrder(){
    System.out.println("Your tree:\n");
    displayInOrder(root);
    System.out.println("\n");
  }

  public void displayInOrder(Node n){
    if(n != null){  
      displayInOrder(n.left);
      System.out.println(n.element);
      displayInOrder(n.right);  
    }
  }

  public void displayPreOrder(){
    System.out.println("Your tree:\n");
    displayPreOrder(root);
    System.out.println("\n");
  }

  private void displayPreOrder(Node n){
    if(n != null){
      System.out.println(n.element);
      displayPreOrder(n.left);
      displayPreOrder(n.right);
    }
  }

  public void displayPostOrder(){
    System.out.println("Your tree: \n");
    displayPostOrder(root); 
    System.out.println("\n");
  }

  private void displayPostOrder(Node n){
    if(n != null){
      displayPostOrder(n.left);
      displayPostOrder(n.right);
      System.out.println(n.element);
    } 
  }

  public Node rotateRight(Node n){
    Node leftSon = n.left;
    n.left = leftSon.right;
    leftSon.right = n;
    return leftSon;
  }

  public Node rotateLeft(Node n){
    Node rightSon = n.right;
    n.right = rightSon.left;
    rightSon.left = n;
    return rightSon;
  }

  public void balance(){

  }
  
  public Node rotateRightLeft(Node n){
    n.right = rotateRight(n.right);
    return rotateLeft(n);
  }

  public Node rotateLeftRight(Node n){
    n.left = rotateLeft(n.left);
    return rotateRight(n);
  }
}
